### PR TITLE
[FM v0.8.0]--Add doc for VPA

### PR DESCRIPTION
**Pull request by jiangpengcheng**
### Motivation

FM v0.8.0 supports VPA ([code PR](https://github.com/streamnative/function-mesh/pull/479)). Therefore, update the doc accordingly.

### Modification
- Update Functions/Source/Sink CRD configurations
- Update the Scaling document

### CODE REVIEW

File: docs/scaling.md [hunk 1]
```diff
@@ -215,13 +221,156 @@ Function Mesh supports automatically scaling up the number of Pods based on a cu
...
+
+- [1] `replicas`: it should be >= a minimal value(default 2) to make VPA can update the pod and ensure that at least (minimalValue - 1) pod is alive during resource updating, you can change the minimal value in `pod.vpa.updatePolicy.minReplicas`.
+- [2] `updateMode`: the update mode defines the behaviour of the VPA Updater, for `Auto` and `Recreate` it will recreate pod when resource recommendation is changed, for `Initial` it will only assigns resources on pod creation, for `Off`, it will do nothing to pod but just save the recommendation resources in VPA object
+- [3] user can specify different resource policy for different containers, `*` means for all containers,
+- [4] the min recommended resources
+- [5] the max recommended resources
+
+##### Enable VPA on requests resources only
+
+Below is an example to update pod request resources.
```
Comment (Huanli-Meng) [inline docs/scaling.md:319]:
```suggestion
Requests define the minimum amount of resources that containers need. This example shows how to update Pods based on resource requests.
```

File: docs/scaling.md [hunk 2]
```diff
@@ -215,13 +221,156 @@ Function Mesh supports automatically scaling up the number of Pods based on a cu
...
+        updateMode: "Auto"
+      resourcePolicy:
+        containerPolicies:
+        - containerName: "*"
+          controlledValues: "RequestsOnly"  --- [1]
+
+  # Other function configs
+```
+
+- [1] the Recommender will only control the **requests** resources, not including **limits** resources, you can change it to `RequestsAndLimits` to control them both
```
Comment (Huanli-Meng) [inline docs/scaling.md:345]:
```suggestion
- [1] `controlledValues`: the VPA Recommender controls the resource **requests**. To control both the requests and limits,  you can set it to `RequestsAndLimits` 
```

File: docs/scaling.md [hunk 3]
```diff
@@ -215,13 +221,156 @@ Function Mesh supports automatically scaling up the number of Pods based on a cu
...
+
+
+##### Enable VPA on cpu resources only
```
Comment (Huanli-Meng) [inline docs/scaling.md:348]:
```suggestion
##### Enable VPA based on CPU resources
```

File: docs/scaling.md [hunk 4]
```diff
@@ -215,13 +221,156 @@ Function Mesh supports automatically scaling up the number of Pods based on a cu
...
+
+Below is an example to update pod request resources.
```
Comment (Huanli-Meng) [inline docs/scaling.md:350]:
```suggestion
This example shows how to update Pods based on CPU resource requests.
```

File: docs/scaling.md [hunk 5]
```diff
@@ -215,13 +221,156 @@ Function Mesh supports automatically scaling up the number of Pods based on a cu
...
+        updateMode: "Auto"
+      resourcePolicy:
+        containerPolicies:
+        - containerName: "*"
+          controlledResources: ["cpu"]   --- [1]
+
+  # Other function configs
+```
+
+- [1] the Recommender will only control the **CPU** resources, you can add **memory** to the array to control it too
```
Comment (Huanli-Meng) [inline docs/scaling.md:376]:
```suggestion
- [1] `controlledResources`: the VPA Recommender controls the **CPU** resources. To control memories, you can add **memory** to the array.
```

**Review by Huanli-Meng — COMMENTED** [review]
Leave my comments. PTAL

File: docs/scaling.md [hunk 6]
```diff
@@ -215,13 +221,156 @@ Function Mesh supports automatically scaling up the number of Pods based on a cu
...
+  1. watches all Pods, keeps calculating fresh recommended resources for them and stores the recommendations in the VPA objects.
+
+  2. exposes a synchronous API that takes a Pod description and returns recommended resources.
+
+- the VPA Admission Controller controls all pod creation requests, If the Pod is matched by any VerticalPodAutoscaler object, the admission controller overrides resources of containers in the Pod with the recommendation provided by the VPA Recommender. If the Recommender is not available, it falls back to the recommendation cached in the VPA object.
+
+- the VPA Updater is a component responsible for real-time updates of Pods. If a Pod uses VPA in "Auto" mode, the Updater can decide to update it with recommender resources.
+
+
+![Architecture](./assets/vpa-architecture.png)
```
[the remark this answers is not in the document]
Comment (jiangpengcheng) [reply docs/scaling.md:254]: this one comes from the official repo of VPA: https://github.com/kubernetes/design-proposals-archive/blob/main/autoscaling/vertical-pod-autoscaler.md#architecture-overview

File: docs/functions/function-crd.md [hunk 7]
```diff
@@ -371,6 +371,10 @@ Function Mesh supports customizing the Pod running function instance. This table
     <td><code>autoScalingBehavior</code></td>
     <td>Configure the scaling behavior of the target in both up and down directions (<code>scaleUp</code> and <code>scaleDown</code> fields respectively). If not specified, the default Kubernetes scaling behaviors are adopted. For details, see <a href="https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#horizontalpodautoscalerbehavior-v2beta2-autoscaling">HorizontalPodAutoscalerBehavior v2beta2 autoscaling</a>. </td>
   </tr>
+  <tr>
+    <td><code>vpa</code></td>
+    <td>Configure the behavior of the vertical pod autoscaling, it contains two fields: <ul> <li><a href="https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-0.12.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L109-L120">updatePolicy</a></li> <li><a href="https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-0.12.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L149-L155">resourcePolicy</a></li> </ul></td>
```
[the remark this answers is not in the document]
Comment (jiangpengcheng) [reply docs/functions/function-crd.md]: ok

File: docs/scaling.md [hunk 8]
```diff
@@ -215,13 +221,156 @@ Function Mesh supports automatically scaling up the number of Pods based on a cu
...
+
+#### How it works
+
+With Kubernetes [Vertical Pod Autoscaler (VPA)](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler), Function Mesh supports automatically update the requested resources(memory and cpu) of Pods (Pulsar instances) that are required to run Pulsar functions, sources, and sinks based on analysis of historical resource utilization, amount of resources available in the cluster and real-time events, such as OOMs.
+
+- the Recommender is a new component which consumes utilization signals and OOM events for all Pods in the cluster from the Metrics Server, it:
+
+  1. watches all Pods, keeps calculating fresh recommended resources for them and stores the recommendations in the VPA objects.
+
+  2. exposes a synchronous API that takes a Pod description and returns recommended resources.
```
[the remark this answers is not in the document]
Comment (jiangpengcheng) [reply docs/scaling.md:247]: hm, they are copied from the official readme of VPA, so I think it's ok to keep it?

**Review by Huanli-Meng — COMMENTED** [review]
leave some comments

**Comment by Huanli-Meng** [comment]
Ping @tpiperatgod or @freeznet for tech review.

**Review by tpiperatgod — APPROVED** [review]

**Review by freeznet — APPROVED** [review]

**Pull request merged**